File: python_core/src/serializer.py

import json
import struct
from typing import Dict, List, Any

class BastionSerializer:
    """
    Enforces 'Deterministic-but-Unique' output signatures.
    """

    ORDER_ROOT = [
        "version", "entropy", "flags", "lastModified", 
        "locker", "contacts", "notes", "configs"
    ]

    ORDER_CONFIG = [
        "id", "name", "username", "category", "version", 
        "length", "useSymbols", "customPassword", "breachStats", "compromised",
        "createdAt", "updatedAt", "usageCount", "sortOrder"
    ]

    ORDER_NOTE = ["id", "updatedAt", "title", "content"]
    ORDER_CONTACT = ["id", "updatedAt", "name", "email", "phone", "address", "notes"]
    ORDER_RESONANCE = ["id", "timestamp", "label", "size", "mime", "key", "hash", "embedded"]

    @staticmethod
    def _reorder(obj: Dict[str, Any], order: List[str]) -> Dict[str, Any]:
        if not isinstance(obj, dict): return obj
        out = {}
        for key in order:
            if key in obj: out[key] = obj[key]
        remaining = sorted([k for k in obj.keys() if k not in order])
        for key in remaining: out[key] = obj[key]
        return out

    @classmethod
    def serialize(cls, state: Any) -> str:
        if hasattr(state, '__dict__'): state_dict = state.__dict__
        elif isinstance(state, dict): state_dict = state
        else: raise ValueError("State must be a dict or dataclass")

        ordered = cls._reorder(state_dict, cls.ORDER_ROOT)

        if 'configs' in ordered and isinstance(ordered['configs'], list):
            ordered['configs'] = [cls._reorder(c, cls.ORDER_CONFIG) for c in ordered['configs']]
        if 'notes' in ordered and isinstance(ordered['notes'], list):
            ordered['notes'] = [cls._reorder(n, cls.ORDER_NOTE) for n in ordered['notes']]
        if 'contacts' in ordered and isinstance(ordered['contacts'], list):
            ordered['contacts'] = [cls._reorder(c, cls.ORDER_CONTACT) for c in ordered['contacts']]
        if 'locker' in ordered and isinstance(ordered['locker'], list):
            ordered['locker'] = [cls._reorder(l, cls.ORDER_RESONANCE) for l in ordered['locker']]

        return json.dumps(ordered, separators=(',', ':'))

    @staticmethod
    def frame(json_str: str) -> bytes:
        """
        Wraps JSON with 4-byte Length Header + 0x00 Padding to 64-byte alignment.
        """
        data_bytes = json_str.encode('utf-8')
        length = len(data_bytes)
        
        # 1. Header
        header = struct.pack('<I', length)
        
        # 2. Padding Calc
        total_raw = 4 + length
        remainder = total_raw % 64
        padding_needed = 0 if remainder == 0 else 64 - remainder
        padding = b'\x00' * padding_needed
        
        return header + data_bytes + padding

    @staticmethod
    def deframe(data: bytes) -> str:
        if len(data) < 4: return data.decode('utf-8')

        try:
            # Read Length (Little Endian)
            claimed_len = struct.unpack('<I', data[:4])[0]
            
            # Check if claimed length fits in buffer (allowing for padding)
            if claimed_len <= len(data) - 4:
                # Valid Bastion Frame: Slice exact payload
                return data[4:4+claimed_len].decode('utf-8')
        except:
            pass
            
        # Fallback
        return data.decode('utf-8')

File: services/serializer.ts

import { VaultState } from "../types";

/**
 * BASTION CANONICAL SERIALIZER (Protocol V3.5)
 * 
 * Enforces "Deterministic-but-Unique" output signatures.
 * 1. Fields are serialized in a strict, non-alphabetical order.
 * 2. Payloads are framed with explicit byte-length prefixes.
 * 3. Total size is aligned to 64-byte blocks via deterministic padding.
 */
export class BastionSerializer {

    // --- 1. FIELD ORDERING DEFINITIONS ---
    
    private static ORDER_ROOT = [
        "version",      // 1. Protocol Version
        "entropy",      // 2. Master Seed
        "flags",        // 3. Feature Flags
        "lastModified", // 4. Timestamp
        "locker",       // 5. Heavy Data
        "contacts",     // 6. Network
        "notes",        // 7. Knowledge
        "configs"       // 8. Credentials
    ];

    private static ORDER_CONFIG = [
        "id", "name", "username", "category", "version", 
        "length", "useSymbols", "customPassword", "breachStats", "compromised",
        "createdAt", "updatedAt", "usageCount", "sortOrder"
    ];

    private static ORDER_NOTE = [
        "id", "updatedAt", "title", "content"
    ];

    private static ORDER_CONTACT = [
        "id", "updatedAt", "name", "email", "phone", "address", "notes"
    ];

    private static ORDER_RESONANCE = [
        "id", "timestamp", "label", "size", "mime", "key", "hash", "embedded"
    ];

    /**
     * Recursively re-orders the VaultState to match the Canonical Signature.
     */
    static serialize(state: VaultState): string {
        const ordered: any = this.reorder(state, this.ORDER_ROOT);

        if (state.configs) ordered.configs = state.configs.map((c: any) => this.reorder(c, this.ORDER_CONFIG));
        if (state.notes) ordered.notes = state.notes.map((n: any) => this.reorder(n, this.ORDER_NOTE));
        if (state.contacts) ordered.contacts = state.contacts.map((c: any) => this.reorder(c, this.ORDER_CONTACT));
        if (state.locker) ordered.locker = state.locker.map((l: any) => this.reorder(l, this.ORDER_RESONANCE));

        return JSON.stringify(ordered);
    }

    private static reorder(obj: any, order: string[]): any {
        if (!obj) return obj;
        const out: any = {};
        
        // 1. Known keys in Strict Order
        for (const key of order) {
            if (Object.prototype.hasOwnProperty.call(obj, key) && obj[key] !== undefined) {
                out[key] = obj[key];
            }
        }
        
        // 2. Unknown keys (Forward Compatibility) - Sorted Alphabetically
        const remaining = Object.keys(obj).filter(k => !order.includes(k)).sort();
        for (const key of remaining) {
             if (obj[key] !== undefined) {
                out[key] = obj[key];
            }
        }
        return out;
    }

    /**
     * Wraps the JSON string in a binary frame with Deterministic Padding.
     * Format: [LENGTH (4B LE)] + [JSON] + [PADDING (0x00...)]
     * Total size aligned to 64 bytes.
     */
    static frame(json: string): Uint8Array {
        const encoder = new TextEncoder();
        const jsonBytes = encoder.encode(json);
        const len = jsonBytes.length;
        
        // Calculate Padding
        // Header (4) + Payload (len) + Padding (p) = Multiple of 64
        const totalRaw = 4 + len;
        const remainder = totalRaw % 64;
        const paddingNeeded = remainder === 0 ? 0 : 64 - remainder;
        
        const buffer = new Uint8Array(totalRaw + paddingNeeded);
        const view = new DataView(buffer.buffer);
        
        // 1. Write Length Header (Little Endian)
        view.setUint32(0, len, true); 
        
        // 2. Write Payload
        buffer.set(jsonBytes, 4);
        
        // 3. Write Padding (Already 0x00 initialized by new Uint8Array)
        if (paddingNeeded > 0) {
            buffer.fill(0, totalRaw, totalRaw + paddingNeeded);
        }
        
        return buffer;
    }

    /**
     * Unwraps a binary frame.
     * STRICT: Reads 4 bytes length, slices exactly that amount.
     * TOLERANT: Ignores trailing bytes (padding).
     */
    static deframe(data: Uint8Array): string {
        if (data.length < 4) {
            // Fallback for malformed V4 blobs (should not happen)
            return new TextDecoder().decode(data);
        }
        
        const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
        const claimedLen = view.getUint32(0, true);
        
        // VALIDATION: Claimed length must fit inside the buffer
        if (claimedLen <= data.length - 4) {
            // Valid Bastion Frame (Slice exact payload, ignore padding)
            return new TextDecoder().decode(data.slice(4, 4 + claimedLen));
        } else {
            // Fallback for Legacy V1/V2 (Raw JSON) if header logic failed
            return new TextDecoder().decode(data);
        }
    }
}
